### ISSUE TITLE
Nano: light weight hpo support in nano pytorch

### ISSUE THREAD

**Issue by shane-huang** [intel-analytics/BigDL#3925]
## Overview

Refer to hpo design in nano tensorflow in issue #3712 

HPO search needs to support below usage scenarios
- torch data loaders (with batch size specs), lightning data modules
- pytorch models, 
- pytorch lightning modules

### ```trainer.search``` 

```python

@automl.obj()
class SegModel(pl.LightningModule):
    def __init__(self,
        data_path: str,
        batch_size: int = 4,
        lr: float = 1e-3,
        num_layers: int = 3,
        features_start: int = 64,
        bilinear: bool = False,
        *args,
        **kwargs
    ):
   ...
 
model = SeqModel(..., lr=space.Real(1e-3,1e-4,log=True), num_layers=space.Categorical[3,6,9], bilinear=space.Bool())
model = trainer.compile(model, ...)
model = trainer.search(model, n_trials=50, ...)
trainer.fit(model)
```
* ```search```return an AutoModel, and `fit` takes an AutoModel or Model. 
* by default the returned AutoModel use the best hparams in search results to instantiate the model. User can use `trainer.end_search` to specify a trial id that he wants to use for `fit`. In this case automodel will use the hparams of the specific trial id. 
* ```search``` **DO NOT automatically search for params like batch size and learning rate**, as pytorch lightening provides ```trainer.tune``` for such purposes (as well as Trainer params `auto_scale_batch_size` and `auto_lr_find`).  If we want to search for bs and lr, we need to align it with the behavior of pytroch tuning.
* ```search``` can be called several times in order to resume the tuning.
* when tuning finished, user can explicitly call ```end_search``` and specify the parameters in which trial will be used for the following fit. If trial id is not specified in fit, the best trial is selected by default. If ```end_search``` is not explicit called, it will be called in ```fit```.
* We assume that it is common to use a subset of data in ```search```, while in ```fit``` the user can use the full datast (train+val used in search), or use a larger dataset, or more epochs.

### ```trainer.search_summary```
```pytorch
...
model = trainer.search(model, n_trials=50, ...)
trainer.search_summary(...)
trainer.fit(model)
```
* ```search_summary``` returns a data frame containing the hparams used in each trial and the value of target metric (and possibly speed or other attributes). Can serve as some sort of a leaderboard.
* search summaries can also be visualized, using tools provided by optuna

## Details and Internals
* Essentially, we support specifying search spaces in :
   * User defined pytorch model (i.e. nn.Module), LightningModule, and LightningDataModule.  User just need to decorate their class using ```automl.obj()```. Then he will be able to specify the search spaces in class arguments. 
  * torch data loaders.  replace ```torch.utils.data.DataLoader``` with ```nano.pytorch.utils.data.DataLoader```. This is automatically gets decorated ```torch.utils.data.DataLoader``` so that parameters like batch size can be specified using search space. 
* ```trainer.compile```  will convert decorated pytorch model to decorated LightingDataModule. ```trainer.compile``` also accepts ```nano.pytroch.optimizers``` which are decorated torch optimizers so that learning rate can also be searched. 
* ```trainer.search``` recognizes all decorocated objects in its arguments. In search the decorated objects are lazily instantiated until in each trial. 
* When model and data are returned from search results, they will be configured with the proper hyperparams and should be a regular LightningModule, LightningDataModule, or DataLoader. 

Will add some exmaples to demostrate the above usages of pytroch models (with optimizers), LightingModules, torch DataLoaders, and LightningDataModule

## Potential issues and considerations 
1. what if the data loaders is different in fit and in search? The batch size returned is the best for data loaders used in search, but is it also good  for data loaders in fit? And if user already specifies a batch size in fit data loaders, shall we silently change the batch size to the batch size obtained from search?
2. Need to consider the use scenario when user does not use automl at all. 

@jason-dai

**Comment by jason-dai** [comment]
We should keep the APIs between Keras and PyTorch as consistent as possible:

1) Use `@nano.automl` instead of `@automl.obj`

2) `Trainer.search` should search the lr, bs, etc. by default, just like `Model.search`

3) I don't think we need `get_search_model`; `Trainer.fit` can automatically obtain the model if it receives an AutoModel

4) How does the user tune each layer in PyTorch (similar to layers in Keras)?

**Comment by shane-huang** [comment]
> We should keep the APIs between Keras and PyTorch as consistent as possible:
> 
> 1. Use `@nano.automl` instead of `@automl.obj`
> 2. `Trainer.search` should search the lr, bs, etc. by default, just like `Model.search`
> 3. I don't think we need `get_search_model`; `Trainer.fit` can automatically obtain the model if it receives an AutoModel
> 4. How does the user tune each layer in PyTorch (similar to layers in Keras)?

1. There're two auto types - **_AutoModels_** and **_AutoObjects_**. In Keras design, layers and optimizers are AutoObjects (which are created implictly inside nano), Customized models, nano.Sequential/Model are AutoModels, which has search, search_summary methods,etc.). In PyTorch since it is trainer that has search/search_summary, models are just AutoObjects, so I decorate it as an `automl.obj()` (In keras a user defined model is decorated using `automl.model()`.  Well I can add another decorator to determine whether to create an AutoModel or AutoObject based on the class type.
 
2. if trainer.search can search lr and bs, how to align it with the behavior of `auto_scale_batch_size` and `auto_lr_find` the in Trainer params. 
3.  How does trainer need a model for fit, whether it is an AutoModel or Model. How should it get it if Trainer.search returns nothing. Do you mean Trainer.search always returns a model? What if user want to instantiate a model from a specific trial id? How should he do it?
4. Tuning of each layer is the same as customized model in Keras. Do you mean we need to support pytorch model creation like Keras Sequential style?

**Comment by shane-huang** [comment]
The implementation is merged so close it. Later may reopen if new features are added.
